llvm+zig HEAD: current installation of LLVM 12 breaks zig HEAD

### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
❯ brew config
HOMEBREW_VERSION: 3.1.9-8-g6f1398a
ORIGIN: https://github.com/Homebrew/brew
HEAD: 6f1398a5338d851151041c81562da38b62cceff3
Last commit: 6 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 051251736a98e9c97996b053c5b7c301b24a8352
Core tap last commit: 73 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 12.0.5 build 1205
Git: 2.30.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.3.1-x86_64
CLT: 12.5.0.22.9
Xcode: N/A

❯ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc
```


### 

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

Current Homebrew's LLVM 12 installation breaks `zig` HEAD by what seems an issue with mixing shared with static libs. In particular, `zig`'s `CMake` config script (which was not changed in between LLVM releases in a way that would affect this, at least not as far as I am aware) configures the project to link dynamically against LLVM libs, i.e., `-lLLVM-12`, while `libclang*` is found and designated for static linking.

**Tl;dr**

```
> brew install zig --HEAD
```

will produce a broken `zig` compiler.

### What happened (include all command output)?

This has very interesting runtime issues (probably a result of a convoluted network of static initializers as described [here](https://reviews.llvm.org/D61804)) which manifest themselves in `zig` no longer being able to act as a C/C++ compiler:

```
❯ zig cc hello.c
hello.c:1:10: fatal error: cannot open file '/Users/jakubkonka/dev/zig/build/lib/zig/include/stdio.h': No such file or directory
#include <stdio.h>
         ^
1 error generated.
```

Here `zig` was installed with `brew install zig --HEAD`. This error was reported to us upstream by multiple users on multiple occasions and here's the tracking issue https://github.com/ziglang/zig/issues/8860. Interestingly enough, this problem was not present with `llvm@11` as `llvm@11` reported LLVM built with static linking unlike `llvm@12`:

```
❯ /usr/local/opt/llvm@11/bin/llvm-config --libs
-lLLVMXRay -lLLVMWindowsManifest -lLLVMTableGen -lLLVMSymbolize -lLLVMDebugInfoPDB -lLLVMOrcJIT -lLLVMOrcError -lLLVMJITLink -lLLVMObjectYAML -lLLVMMCA -lLLVMLTO -lLLVMObjCARCOpts -lLLVMExtensions -lPolly -lPollyISL -lLLVMPasses -lLLVMCoroutines -lLLVMLineEditor -lLLVMLibDriver -lLLVMInterpreter -lLLVMFuzzMutate -lLLVMMCJIT -lLLVMExecutionEngine -lLLVMRuntimeDyld -lLLVMDWARFLinker -lLLVMDlltoolDriver -lLLVMOption -lLLVMDebugInfoGSYM -lLLVMCoverage -lLLVMXCoreDisassembler -lLLVMXCoreCodeGen -lLLVMXCoreDesc -lLLVMXCoreInfo -lLLVMX86Disassembler -lLLVMX86AsmParser -lLLVMX86CodeGen -lLLVMX86Desc -lLLVMX86Info -lLLVMWebAssemblyDisassembler -lLLVMWebAssemblyCodeGen -lLLVMWebAssemblyDesc -lLLVMWebAssemblyAsmParser -lLLVMWebAssemblyInfo -lLLVMSystemZDisassembler -lLLVMSystemZCodeGen -lLLVMSystemZAsmParser -lLLVMSystemZDesc -lLLVMSystemZInfo -lLLVMSparcDisassembler -lLLVMSparcCodeGen -lLLVMSparcAsmParser -lLLVMSparcDesc -lLLVMSparcInfo -lLLVMRISCVDisassembler -lLLVMRISCVCodeGen -lLLVMRISCVAsmParser -lLLVMRISCVDesc -lLLVMRISCVUtils -lLLVMRISCVInfo -lLLVMPowerPCDisassembler -lLLVMPowerPCCodeGen -lLLVMPowerPCAsmParser -lLLVMPowerPCDesc -lLLVMPowerPCInfo -lLLVMNVPTXCodeGen -lLLVMNVPTXDesc -lLLVMNVPTXInfo -lLLVMMSP430Disassembler -lLLVMMSP430CodeGen -lLLVMMSP430AsmParser -lLLVMMSP430Desc -lLLVMMSP430Info -lLLVMMipsDisassembler -lLLVMMipsCodeGen -lLLVMMipsAsmParser -lLLVMMipsDesc -lLLVMMipsInfo -lLLVMLanaiDisassembler -lLLVMLanaiCodeGen -lLLVMLanaiAsmParser -lLLVMLanaiDesc -lLLVMLanaiInfo -lLLVMHexagonDisassembler -lLLVMHexagonCodeGen -lLLVMHexagonAsmParser -lLLVMHexagonDesc -lLLVMHexagonInfo -lLLVMBPFDisassembler -lLLVMBPFCodeGen -lLLVMBPFAsmParser -lLLVMBPFDesc -lLLVMBPFInfo -lLLVMAVRDisassembler -lLLVMAVRCodeGen -lLLVMAVRAsmParser -lLLVMAVRDesc -lLLVMAVRInfo -lLLVMARMDisassembler -lLLVMARMCodeGen -lLLVMARMAsmParser -lLLVMARMDesc -lLLVMARMUtils -lLLVMARMInfo -lLLVMAMDGPUDisassembler -lLLVMAMDGPUCodeGen -lLLVMMIRParser -lLLVMipo -lLLVMInstrumentation -lLLVMVectorize -lLLVMLinker -lLLVMIRReader -lLLVMAsmParser -lLLVMFrontendOpenMP -lLLVMAMDGPUAsmParser -lLLVMAMDGPUDesc -lLLVMAMDGPUUtils -lLLVMAMDGPUInfo -lLLVMAArch64Disassembler -lLLVMMCDisassembler -lLLVMAArch64CodeGen -lLLVMCFGuard -lLLVMGlobalISel -lLLVMSelectionDAG -lLLVMAsmPrinter -lLLVMDebugInfoDWARF -lLLVMCodeGen -lLLVMTarget -lLLVMScalarOpts -lLLVMInstCombine -lLLVMAggressiveInstCombine -lLLVMTransformUtils -lLLVMBitWriter -lLLVMAnalysis -lLLVMProfileData -lLLVMObject -lLLVMTextAPI -lLLVMBitReader -lLLVMCore -lLLVMRemarks -lLLVMBitstreamReader -lLLVMAArch64AsmParser -lLLVMMCParser -lLLVMAArch64Desc -lLLVMMC -lLLVMDebugInfoCodeView -lLLVMDebugInfoMSF -lLLVMBinaryFormat -lLLVMAArch64Utils -lLLVMAArch64Info -lLLVMSupport -lLLVMDemangle

❯ /usr/local/opt/llvm@12/bin/llvm-config --libs
-lLLVM-12
```

**Tl;dr**

Installing `zig` HEAD and running tests will fail:

```
> brew install zig --HEAD
> brew test zig
==> Testing zig
==> /usr/local/Cellar/zig/HEAD-2f538f3_1/bin/zig build-exe hello.zig
==> ./hello
==> /usr/local/Cellar/zig/HEAD-2f538f3_1/bin/zig cc hello.c -o hello
Last 15 lines from /Users/jakubkonka/Library/Logs/Homebrew/zig/test.02.zig:
2021-05-25 23:16:22 +0200

/usr/local/Cellar/zig/HEAD-2f538f3_1/bin/zig
cc
hello.c
-o
hello

hello.c:1:10: fatal error: cannot open file '/usr/local/include/stdio.h': No such file or directory
#include <stdio.h>
         ^
1 error generated.
Error: zig: failed
An exception occurred within a child process:
  BuildError: Failed executing: /usr/local/Cellar/zig/HEAD-2f538f3_1/bin/zig cc hello.c -o hello
/usr/local/Homebrew/Library/Homebrew/formula.rb:2175:in `block in system'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2111:in `open'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2111:in `system'
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/zig.rb:49:in `block in <class:Zig>'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1975:in `block (3 levels) in run_test'
/usr/local/Homebrew/Library/Homebrew/utils.rb:558:in `with_env'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1974:in `block (2 levels) in run_test'
/usr/local/Homebrew/Library/Homebrew/formula.rb:924:in `with_logging'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1973:in `block in run_test'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `block in run'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `chdir'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:63:in `run'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2224:in `mktemp'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1967:in `run_test'
/usr/local/Homebrew/Library/Homebrew/test.rb:43:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `block in catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:33:in `catch'
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/timeout.rb:108:in `timeout'
/usr/local/Homebrew/Library/Homebrew/test.rb:48:in `<main>'
```

### What did you expect to happen?

The tests should pass for `brew install zig --HEAD`.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
> brew install zig --HEAD
> brew test zig
```

Unfortunately, I've just discovered that adding either `ZIG_STATIC_LLVM` or `ZIG_PREFER_CLANG_CPP_DYLIB` to CMake flags in Homebrew (or in general to CMake when linking against Homebrew's built LLVM12) still won't cut. While the smoke tests I added to the zig formula pass just fine with this tweak, the built compiler later fails at building stage2 tests:

```
$ brew test zig
==> Testing zig
==> /usr/local/Cellar/zig/HEAD-d8d92da_1/bin/zig build-exe hello.zig
==> ./hello
==> /usr/local/Cellar/zig/HEAD-d8d92da_1/bin/zig cc hello.c -o hello
==> ./hello

$ zig build test-stage2 -Denable-llvm
error(compilation): clang failed with stderr: /Users/jakubkonka/dev/zig/src/zig_clang.cpp:22:10: fatal error: 'clang/Frontend/ASTUnit.h' file not found

error(compilation): clang failed with stderr: /Users/jakubkonka/dev/zig/src/zig_clang_cc1as_main.cpp:14:10: fatal error: 'clang/Basic/Diagnostic.h' file not found

error(compilation): clang failed with stderr: /Users/jakubkonka/dev/zig/src/zig_clang_cc1_main.cpp:15:10: fatal error: 'clang/Basic/Stack.h' file not found

error(compilation): clang failed with stderr: In file included from /Users/jakubkonka/dev/zig/src/zig_llvm.cpp:16:
/Users/jakubkonka/dev/zig/src/zig_llvm.h:13:10: fatal error: 'llvm-c/Core.h' file not found

error(compilation): clang failed with stderr: /Users/jakubkonka/dev/zig/src/zig_clang_driver.cpp:14:10: fatal error: 'clang/Driver/Driver.h' file not found

/Users/jakubkonka/dev/zig/src/zig_clang.cpp:1:1: error: unable to build C object: clang exited with code 1
/Users/jakubkonka/dev/zig/src/zig_clang_cc1as_main.cpp:1:1: error: unable to build C object: clang exited with code 1
/Users/jakubkonka/dev/zig/src/zig_llvm.cpp:1:1: error: unable to build C object: clang exited with code 1
/Users/jakubkonka/dev/zig/src/zig_clang_driver.cpp:1:1: error: unable to build C object: clang exited with code 1
/Users/jakubkonka/dev/zig/src/zig_clang_cc1_main.cpp:1:1: error: unable to build C object: clang exited with code 1
The following command exited with error code 1:
/usr/local/Cellar/zig/HEAD-d8d92da_1/bin/zig test /Users/jakubkonka/dev/zig/src/test.zig -cflags -std=c++14 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D_GNU_SOURCE -fvisibility-inlines-hidden -fno-exceptions -fno-rtti -Werror=type-limits -Wno-missing-braces -Wno-comment -DNDEBUG=1 -- /Users/jakubkonka/dev/zig/src/zig_llvm.cpp /Users/jakubkonka/dev/zig/src/zig_clang.cpp /Users/jakubkonka/dev/zig/src/zig_clang_driver.cpp /Users/jakubkonka/dev/zig/src/zig_clang_cc1_main.cpp /Users/jakubkonka/dev/zig/src/zig_clang_cc1as_main.cpp /Users/jakubkonka/dev/zig/src/windows_sdk.cpp -lclangFrontendTool -lclangCodeGen -lclangFrontend -lclangDriver -lclangSerialization -lclangSema -lclangStaticAnalyzerFrontend -lclangStaticAnalyzerCheckers -lclangStaticAnalyzerCore -lclangAnalysis -lclangASTMatchers -lclangAST -lclangParse -lclangSema -lclangBasic -lclangEdit -lclangLex -lclangARCMigrate -lclangRewriteFrontend -lclangRewrite -lclangCrossTU -lclangIndex -lclangToolingCore -llldDriver -llldMinGW -llldELF -llldCOFF -llldMachO -llldWasm -llldReaderWriter -llldCore -llldYAML -llldCommon -lLLVMWindowsManifest -lLLVMXRay -lLLVMLibDriver -lLLVMDlltoolDriver -lLLVMCoverage -lLLVMLineEditor -lLLVMXCoreDisassembler -lLLVMXCoreCodeGen -lLLVMXCoreDesc -lLLVMXCoreInfo -lLLVMX86Disassembler -lLLVMX86AsmParser -lLLVMX86CodeGen -lLLVMX86Desc -lLLVMX86Info -lLLVMWebAssemblyDisassembler -lLLVMWebAssemblyAsmParser -lLLVMWebAssemblyCodeGen -lLLVMWebAssemblyDesc -lLLVMWebAssemblyInfo -lLLVMSystemZDisassembler -lLLVMSystemZAsmParser -lLLVMSystemZCodeGen -lLLVMSystemZDesc -lLLVMSystemZInfo -lLLVMSparcDisassembler -lLLVMSparcAsmParser -lLLVMSparcCodeGen -lLLVMSparcDesc -lLLVMSparcInfo -lLLVMRISCVDisassembler -lLLVMRISCVAsmParser -lLLVMRISCVCodeGen -lLLVMRISCVDesc -lLLVMRISCVInfo -lLLVMPowerPCDisassembler -lLLVMPowerPCAsmParser -lLLVMPowerPCCodeGen -lLLVMPowerPCDesc -lLLVMPowerPCInfo -lLLVMNVPTXCodeGen -lLLVMNVPTXDesc -lLLVMNVPTXInfo -lLLVMMSP430Disassembler -lLLVMMSP430AsmParser -lLLVMMSP430CodeGen -lLLVMMSP430Desc -lLLVMMSP430Info -lLLVMMipsDisassembler -lLLVMMipsAsmParser -lLLVMMipsCodeGen -lLLVMMipsDesc -lLLVMMipsInfo -lLLVMLanaiDisassembler -lLLVMLanaiCodeGen -lLLVMLanaiAsmParser -lLLVMLanaiDesc -lLLVMLanaiInfo -lLLVMHexagonDisassembler -lLLVMHexagonCodeGen -lLLVMHexagonAsmParser -lLLVMHexagonDesc -lLLVMHexagonInfo -lLLVMBPFDisassembler -lLLVMBPFAsmParser -lLLVMBPFCodeGen -lLLVMBPFDesc -lLLVMBPFInfo -lLLVMAVRDisassembler -lLLVMAVRAsmParser -lLLVMAVRCodeGen -lLLVMAVRDesc -lLLVMAVRInfo -lLLVMARMDisassembler -lLLVMARMAsmParser -lLLVMARMCodeGen -lLLVMARMDesc -lLLVMARMUtils -lLLVMARMInfo -lLLVMAMDGPUDisassembler -lLLVMAMDGPUAsmParser -lLLVMAMDGPUCodeGen -lLLVMAMDGPUDesc -lLLVMAMDGPUUtils -lLLVMAMDGPUInfo -lLLVMAArch64Disassembler -lLLVMAArch64AsmParser -lLLVMAArch64CodeGen -lLLVMAArch64Desc -lLLVMAArch64Utils -lLLVMAArch64Info -lLLVMOrcJIT -lLLVMMCJIT -lLLVMJITLink -lLLVMOrcTargetProcess -lLLVMOrcShared -lLLVMInterpreter -lLLVMExecutionEngine -lLLVMRuntimeDyld -lLLVMSymbolize -lLLVMDebugInfoPDB -lLLVMDebugInfoGSYM -lLLVMOption -lLLVMObjectYAML -lLLVMMCA -lLLVMMCDisassembler -lLLVMLTO -lLLVMPasses -lLLVMCFGuard -lLLVMCoroutines -lLLVMObjCARCOpts -lLLVMHelloNew -lLLVMipo -lLLVMVectorize -lLLVMLinker -lLLVMInstrumentation -lLLVMFrontendOpenMP -lLLVMFrontendOpenACC -lLLVMExtensions -lLLVMDWARFLinker -lLLVMGlobalISel -lLLVMMIRParser -lLLVMAsmPrinter -lLLVMDebugInfoDWARF -lLLVMSelectionDAG -lLLVMCodeGen -lLLVMIRReader -lLLVMAsmParser -lLLVMInterfaceStub -lLLVMFileCheck -lLLVMFuzzMutate -lLLVMTarget -lLLVMScalarOpts -lLLVMInstCombine -lLLVMAggressiveInstCombine -lLLVMTransformUtils -lLLVMBitWriter -lLLVMAnalysis -lLLVMProfileData -lLLVMObject -lLLVMTextAPI -lLLVMMCParser -lLLVMMC -lLLVMDebugInfoCodeView -lLLVMDebugInfoMSF -lLLVMBitReader -lLLVMCore -lLLVMRemarks -lLLVMBitstreamReader -lLLVMBinaryFormat -lLLVMSupport -lLLVMDemangle -lc++ -lc --pkg-begin build_options /Users/jakubkonka/dev/zig/zig-cache/test_build_options.zig --pkg-end --cache-dir /Users/jakubkonka/dev/zig/zig-cache --global-cache-dir /Users/jakubkonka/.cache/zig --name test --pkg-begin stage2_tests /Users/jakubkonka/dev/zig/test/stage2/test.zig --pkg-end
error: the following build command failed with exit code 1:
/Users/jakubkonka/dev/zig/zig-cache/o/638aabbbae2995a29c3549f94d117c93/build /usr/local/Cellar/zig/HEAD-d8d92da_1/bin/zig /Users/jakubkonka/dev/zig /Users/jakubkonka/dev/zig/zig-cache /Users/jakubkonka/.cache/zig test-stage2 -Denable-llvm
```

--- a/Formula/zig.rb
+++ b/Formula/zig.rb
@@ -24,7 +24,7 @@ class Zig < Formula
   depends_on "cmake" => :build
 
   def install
-    system "cmake", ".", *std_cmake_args
+    system "cmake", ".", *std_cmake_args, "-DZIG_STATIC_LLVM=ON"
     system "make", "install"
   end
 

